engine/base_client/search.py

```diff
@@ -47,11 +47,22 @@ def _search_one(cls, query: Query, top: Optional[int] = None):
         end = time.perf_counter()
 
         precision = 1.0
+        recall = 1.0
         if query.expected_result:
             ids = set(x[0] for x in search_res)
-            precision = len(ids.intersection(query.expected_result[:top])) / top
 
-        return precision, end - start
+            # precision equals True Positives / (True Positives + False Positives)
+            # recall equals True Positives / (True Positives + False Negatives)
+            # See https://en.wikipedia.org/wiki/Precision_and_recall
+            true_positives = len(ids.intersection(query.expected_result[:top]))
+            precision = (
+                true_positives / top
+            )  # 1 means that the results consist of expected elements.
+            recall = true_positives / len(
+                query.expected_result
+            )  # 1 means that all expected elements are in the results.
+
+        return precision, recall, end - start
```

Comment:
Maybe returning a data structure for all metrics for instance : `{ "precision":.., "recall":... }` might be more suitable and flexible as it do not change the output of the function. Also, it seems that here, precision and recall are the same.``

Reply:
Taking into account that in line 39 we have

```
top = (
                len(query.expected_result)
                if query.expected_result is not None and len(query.expected_result) > 0
                else DEFAULT_TOP
            )
```

Is there any difference in recall and precision if `top` is not specified?